Mathlib/RingTheory/SurjectiveOnStalks.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Localization.AtPrime
+import Mathlib.RingTheory.TensorProduct.Basic
+
+/-!
+# Ring Homomorphisms surjective on stalks
+
+In this file, we prove some results on ring homomorphisms surjective on stalks, to be used in
+the development of immersions in algebraic geometry.
+
+-/
+
+variable {R} [CommRing R] (M : Submonoid R) {S} [CommRing S] [Algebra R S]
+variable {T} [CommRing T] [Algebra R T] [Algebra S T] [IsScalarTower R S T]
+variable {g : S →+* T} {f : R →+* S}
+
+namespace RingHom
+
+/--
+A ring homomorphism `R →+* S` is surjective on stalks if `R_p →+* S_q` is surjective for all pairs
+of primes `p = f⁻¹(q)`.
+-/
+def SurjectiveOnStalks (f : R →+* S) : Prop :=
+  ∀ (I : Ideal S) (_ : I.IsPrime), Function.Surjective (Localization.localRingHom _ I f rfl)
+
+lemma surjective_localRingHom_iff (P : Ideal S) [P.IsPrime] :
```

Comment:
Maybe this lemma deserves a docstring to explain its usage, since it's used a lot.